### PR TITLE
Nullity annotations for the most used APIs

### PR DESCRIPTION
**Pull request by jreznot**
## Proposed changes

At the moment, IntelliJ IDEA shows this warning in Kotlin because there is no nullity information in Selenide:

image

These changes improves IDE experience by providing exact type of parameters and return values including information on null/notnull values. This also makes Selenide work better with Kotlin as Kotlin now knows types and may omit excessive warnings and runtime checks.

image

## Checklist
- [x] Checkstyle and unit tests are passed locally with my changes by running `gradlew check chrome_headless firefox_headless` command
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

Checkstyle warns about SelenideDriver for some reason.
```
[ant:checkstyle] [ERROR] /home/travis/build/selenide/selenide/src/main/java/com/codeborne/selenide/SelenideDriver.java:32:1: Class Fan-Out Complexity is 30 (max allowed is 29). [ClassFanOutComplexity]
```

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/SelenideElement.java [hunk 1]
```diff
@@ -923,26 +995,27 @@
...
   boolean isImage();
 
   /**
    * Take screenshot of this element
-   * @return file with screenshot (*.png)
    *
+   * @return file with screenshot (*.png)
    * @see com.codeborne.selenide.commands.TakeScreenshot
    */
+  @Nonnull
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/SelenideElement.java:1009]: AFAIK it is Nullable 
see https://github.com/selenide/selenide/blob/master/src/main/java/com/codeborne/selenide/impl/ScreenShotLaboratory.java#L110

File: src/main/java/com/codeborne/selenide/SelenideElement.java [hunk 2]
```diff
@@ -923,26 +995,27 @@
...
 
   /**
    * Take screenshot of this element
-   * @return buffered image with screenshot
    *
+   * @return buffered image with screenshot
    * @see com.codeborne.selenide.commands.TakeScreenshotAsImage
    */
   @CheckReturnValue
+  @Nonnull
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/SelenideElement.java:1019]: AFAIK it is Nullable. The same as for TakeScreenshot

File: src/main/java/com/codeborne/selenide/SelenideElement.java [hunk 3]
```diff
@@ -106,95 +114,114 @@
...
    * Get the property value of the pseudo-element
+   *
    * @param pseudoElementName pseudo-element name of the element,
    *                          ":before", ":after", ":first-letter", ":first-line", ":selection"
-   * @param propertyName property name of the pseudo-element
+   * @param propertyName      property name of the pseudo-element
    * @return the property value or "" if the property is missing
    */
   @CheckReturnValue
+  @Nullable
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/SelenideElement.java:204]: javadoc says `@return the property value or "" if the property is missing`
is it really nullable? if it is true please fix javadoc then

File: src/main/java/com/codeborne/selenide/SelenideDriver.java [hunk 4]
```diff
@@ -300,34 +336,41 @@ public boolean atBottom() {
     return executeJavaScript("return window.pageYOffset + window.innerHeight >= document.body.scrollHeight");
   }
 
+  @Nonnull
   public SelenideTargetLocator switchTo() {
     return driver().switchTo();
   }
 
   @CheckReturnValue
+  @Nullable
```
Comment (BorisOsipov) [inline src/main/java/com/codeborne/selenide/SelenideDriver.java:345]: why is it nullable? Don't see in selenium client javadoc any info https://www.javadoc.io/doc/org.seleniumhq.selenium/selenium-api/3.141.59/org/openqa/selenium/WebDriver.html#getCurrentUrl--

also in w3c spec there is no info about return null
https://www.w3.org/TR/webdriver/#get-current-url

**Comment by BorisOsipov** [comment]
>Checkstyle warns about SelenideDriver for some reason.
>SelenideDriver.java:32:1: Class Fan-Out Complexity is 30 (max allowed is 29). [ClassFanOutComplexity]

In fact it is valid warn - `ClassFanOutComplexity` counts all type references tokens and fails check when total count is more than threshold value.

Seems we have to add `SelenideDriver.java` to exclusion list for ClassFanOutComplexity rule or somehow change exclusion package patterns.

cc: @selenide/committers 

For more details https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.html

**Comment by jreznot** [comment]
I've fixed checktyle config and all return value annotations mentioned in comments

**Comment by asolntsev** [comment]
@jreznot Several chrome and firefox tests are failing:

```
java.lang.IllegalArgumentException: Unknown Selenide method: getAttribute
	at com.codeborne.selenide.commands.Commands.execute(Commands.java:146)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:100)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:66)
	at com.sun.proxy.$Proxy28.getAttribute(Unknown Source)
	at integration.AttributeTest.userCanFindElementByAttribute(AttributeTest.java:53)
```

**Comment by rosolko** [comment]
Now tests throw such exception: 
```java
java.lang.IllegalArgumentException: object is not an instance of declaring class
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:108)
        at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:71)
        at com.sun.proxy.$Proxy29.getAttribute(Unknown Source)
        at com.codeborne.selenide.commands.GetSelectedValue.execute(GetSelectedValue.java:24)
        at com.codeborne.selenide.commands.GetSelectedValue.execute(GetSelectedValue.java:10)
        at com.codeborne.selenide.commands.Commands.execute(Commands.java:148)
        at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:105)
        at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:71)
        at com.sun.proxy.$Proxy29.getSelectedValue(Unknown Source)
        at integration.DynamicSelectsTest.waitsUntilOptionWithValueAppears(DynamicSelectsTest.java:37)

**Comment by jreznot** [comment]
Please do not check it, I am still in progress )

**Comment by jreznot** [comment]
Everything fixed now, Chrome still failed with `Caused by: java.net.ConnectException: Failed to connect to localhost/127.0.0.1:28901`

**Comment by asolntsev** [comment]
> Everything fixed now, Chrome still failed with `Caused by: java.net.ConnectException: Failed to connect to localhost/127.0.0.1:28901`

It was a flaky test: chrome crashes on Travis CI time-to-time.

**Comment by plagov** [comment]
@jreznot, thanks for your contribution to improving Kotlin support in Selenide.

I used to have such code:
```
fun getCustomerName(): String {
    return container.find("[data-testid='customer-name']").text
}
```
After upgrade to 5.12.0 it doesn't compile with:
```
Type mismatch.
Required: String
Found: String?
```
Of course, I can change the return type of the method from `String` to `String?`, but that will require to change all subsequent calls, which I don't want. Or I can do the not-null call, `find("...").text!!`, but that is not a recommended way, as far as I understand.
Can you please advise what will be the proper way to handle this?

**Comment by asolntsev** [comment]
@plagov thank you for the feedback.

@jreznot probably in this case we should change return type to `String` because (I guess) `getText()` never returns `null`.

**Comment by pavelpp** [comment]
I've never seen getText() return null.

**Comment by jreznot** [comment]
@asolntsev I've set text as nullable because `com.codeborne.selenide.commands.GetText` calls `GetSelectedText`for `select` tag and it may return null as it is stated here:
```
public class GetSelectedText implements Command<String> {
  ... 
  @Override
  public String execute(SelenideElement proxy, WebElementSource selectElement, Object[] args) {
    WebElement option = getSelectedOption.execute(proxy, selectElement, NO_ARGS);
    return option == null ? null : option.getText();
  }
}
```

**Comment by jreznot** [comment]
@pavelpp 

> Also, Selenide.executeJavaScript() returns String? now, though I'm not sure it can return null.

WebDriver JavaDoc for `org.openqa.selenium.JavascriptExecutor#executeScript`:
```
   * @return One of Boolean, Long, Double, String, List, Map or WebElement. Or null.
   */
  Object executeScript(String script, Object... args);
```

**Comment by asolntsev** [comment]
1. I think we can change `GetSelectedText` to always return empty string.
2. But `executeJavascript` can return null for sure.

**Comment by jreznot** [comment]
>     1. I think we can change `GetSelectedText` to always return empty string.

@asolntsev Isn't it a breaking change? In this case you cannot distinguish empty text of selected option and no selection in select.

**Comment by pavelpp** [comment]
@jreznot Actually it is not possible to have no selection. It will fail much earierl. `  @Override
  public WebElement getFirstSelectedOption() {
    return getOptions().stream().filter(WebElement::isSelected).findFirst()
        .orElseThrow(() -> new NoSuchElementException("No options are selected"));
  }
`

**Pull request merged**